File: pp-exercises/mpi/exercise4/ring-torus.f90
program main  ! torus topology

  implicit none
  include 'mpif.h'
  integer :: size, rank, rank2, rank3, ierr
  integer :: comm_torus, comm_row
  integer :: partner_left, partner_right
  integer :: sum, sbuf, rbuf, i
  integer :: dims(2), coords(2)
  logical :: periods(2), remain_dims(2)
  character(8) :: arg

  call mpi_init(ierr)
  call mpi_comm_size(MPI_COMM_WORLD, size, ierr)
  call mpi_comm_rank(MPI_COMM_WORLD, rank, ierr)

  if (command_argument_count() /= 2) then
     write(0,*) "Usage: ring-torus dims1 dims2"
     call mpi_abort(mpi_comm_world, 1, ierr)
  endif

  call get_command_argument(1, arg) ;  read(arg, *) dims(1)
  call get_command_argument(2, arg) ;  read(arg, *) dims(2)

  if (dims(1) * dims(2) /= size) then
     write(0,*) "ring-torus: dims1 * dims2 /= n_procs"
     call mpi_abort(mpi_comm_world, 1, ierr)
  endif

  periods(1) = .true.
  periods(2) = .true. 
  remain_dims(1) = .false.
  remain_dims(2) = .true.

  call mpi_cart_create(MPI_COMM_WORLD, 2, dims, periods,.true.,comm_torus, ierr)
  call mpi_comm_rank(comm_torus, rank2, ierr)
  call mpi_cart_coords(comm_torus, rank2, 2, coords, ierr)

  call mpi_cart_sub(comm_torus, remain_dims, comm_row, ierr)
  call mpi_comm_rank(comm_row, rank3, ierr)
  
  write(6, "(5(a,i0),a)") &
      "ranks(orig,torus,row) = ", rank, ",", rank2, ",", rank3, &
      " : coords (x,y) = (", coords(1), ",", coords(2), ")" 

  call mpi_cart_shift(comm_row, 0, 1, partner_left, partner_right, ierr)

  sum = 0
  sbuf = rank3
  do 
     call mpi_sendrecv(sbuf, 1, MPI_INTEGER, partner_right, 0, &
                       rbuf, 1, MPI_INTEGER, partner_left, 0, &
                       comm_row, MPI_STATUS_IGNORE, ierr)
     sum = sum + rbuf
     sbuf = rbuf
     if (rbuf == rank3) exit
  enddo  

  write(6,"(3(ai0))") "torus-sum(", coords(1), ",", coords(2), ") = ", sum

  call mpi_finalize(ierr)
end

File: pp-exercises/mpi/exercise4/ring-torus.c
#include <stdio.h>
#include <stdlib.h>
#include <mpi.h>

int main(int argc, char *argv[])  // torus topology
{
    MPI_Comm comm_ring, comm_torus, comm_row;
    int size, rank, rank2, rank3;
    int partner_left, partner_right;
    int sum, sbuf, rbuf, i;
    int dims[2], coords[2], periods[2], remain_dims[2];

    MPI_Init(&argc, &argv);
    MPI_Comm_size(MPI_COMM_WORLD, &size);
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);

    if (argc != 3) {
        fprintf(stderr, "Usage: ring-torus dims1 dims2\n");
        MPI_Abort(MPI_COMM_WORLD, 1);
    }

    dims[0] = atoi(argv[1]);
    dims[1] = atoi(argv[2]);

    if (dims[0] * dims[1] != size) {
	fprintf(stderr, "ring-torus: dims1 * dims2 /= n_procs\n");
	MPI_Abort(MPI_COMM_WORLD, 1);
    }

    periods[0] = 1;
    periods[1] = 1; 
    remain_dims[0] = 0;
    remain_dims[1] = 1;

    MPI_Cart_create(MPI_COMM_WORLD, 2, dims, periods, 1, &comm_torus);
    MPI_Comm_rank(comm_torus, &rank2);
    MPI_Cart_coords(comm_torus, rank2, 2, coords);

    MPI_Cart_sub(comm_torus, remain_dims, &comm_row);
    MPI_Comm_rank(comm_row, &rank3);
  
    printf("ranks(orig,torus,row) = %d,%d,%d : coords (x,y) = (%d,%d)\n", 
	   rank, rank2, rank3, coords[0], coords[1]);

    MPI_Cart_shift(comm_row, 0, 1, &partner_left, &partner_right);

    sum = 0;
    sbuf = rank3;

    do {
	MPI_Sendrecv(&sbuf, 1, MPI_INT, partner_right, 0,
                     &rbuf, 1, MPI_INT, partner_left, 0, 
                     comm_row, MPI_STATUS_IGNORE);
	sum += rbuf;
	sbuf = rbuf;
    } while (rbuf != rank3);

    printf("torus-sum(%d,%d) = %d\n", coords[0], coords[1], sum);

    MPI_Finalize();
    return 0;
}
